[Bug] lazily create locationProvider in SerializableTable for iceberg tables

<!-- (At least include the following, feel free to add if you have more content) -->

### Steps to reproduce the behavior (Required)
1. CREATE TABLE 

```sql
CREATE TABLE IF NOT EXISTS demo.db.table (
  ns_creation_timestamp TIMESTAMP_NTZ,
  ns_tenant_id BIGINT,
  timestamp TIMESTAMP_NTZ,
  alert_type STRING)
USING iceberg
PARTITIONED BY (ns_tenant_id, day(timestamp))
LOCATION 's3://bucket/table'
TBLPROPERTIES (
  'format' = 'iceberg/parquet',
  'format-version' = '2',
  'write.location-provider.impl' = 'com.company.CustomLocationProvider',
  'write.parquet.compression-codec' = 'gzip');
```

2. SELECT  

```sql
set plan_mode = "distributed";
select ns_tenant_id, count(1) from iceberg_catalog.db.table group by ns_tenant_id order by count(1) desc limit 10;
```

### Expected behavior (Required)

Starrocks return the query result correctly without using the custom location provider, which is only useful in the write path. Similar changes have been merged in upstream 
- https://github.com/apache/iceberg/pull/9029
- https://github.com/apache/iceberg/pull/14280

### Real behavior (Required)

Starrocks returns error with the following logs.

```
2026-05-14 06:26:43.702Z INFO (iceberg_glue-ns_pilot_common-alert-1778740003701-fetch_result|113494) [CatalogUtil.loadFileIO():382] Loading custom FileIO implementation: com.starrocks.connector.iceberg.io.IcebergCachingFileIO
2026-05-14 06:26:43.782Z INFO (iceberg_glue-ns_pilot_common-alert-1778740003701-fetch_result|113494) [BaseMetastoreTableOperations.refreshFromMetadataLocation():190] Refreshing table metadata from new version: s3://ns-pilot-common/alert/metadata/16114-46597b12-cc4f-45ac-96f0-25508a75a129.metadata.json
2026-05-14 06:26:43.782Z INFO (iceberg_glue-ns_pilot_common-alert-1778740003701-fetch_result|113494) [CatalogUtil.loadFileIO():382] Loading custom FileIO implementation: org.apache.iceberg.aws.s3.S3FileIO
2026-05-14 06:26:43.911Z INFO (iceberg_glue-ns_pilot_common-alert-1778740003701-fetch_result|113494) [BaseMetastoreCatalog.loadTable():69] Table loaded by catalog: iceberg_glue.ns_pilot_common.alert
2026-05-14 06:26:43.916Z ERROR (iceberg_glue-ns_pilot_common-alert-1778740003701-fetch_result|113494) [MetadataMgr.getSerializedMetaSpec():849] Failed to get remote meta splits on catalog [iceberg_glue], table [ns_pilot_common.alert]
java.lang.IllegalArgumentException: Unable to find a constructor for implementation com.company.CustomLocationProvider of interface org.apache.iceberg.io.LocationProvider. Make sure the implementation is in classpath, and that it either has a public no-arg constructor or a two-arg constructor taking in the string base table location and its property string map.
    at org.apache.iceberg.LocationProviders.locationsFor(LocationProviders.java:53) ~[iceberg-core-1.10.0.jar:?]
    at org.apache.iceberg.BaseMetastoreTableOperations.locationProvider(BaseMetastoreTableOperations.java:234) ~[iceberg-core-1.10.0.jar:?]
    at org.apache.iceberg.BaseTable.locationProvider(BaseTable.java:255) ~[iceberg-core-1.10.0.jar:?]
    at com.starrocks.connector.share.iceberg.SerializableTable.<init>(SerializableTable.java:101) ~[starrocks-hadoop-ext.jar:?]
    at com.starrocks.connector.iceberg.IcebergMetadata.getSerializedMetaSpec(IcebergMetadata.java:634) ~[starrocks-fe.jar:?]
    at com.starrocks.connector.CatalogConnectorMetadata.getSerializedMetaSpec(CatalogConnectorMetadata.java:193) ~[starrocks-fe.jar:?]
    at com.starrocks.server.MetadataMgr.getSerializedMetaSpec(MetadataMgr.java:847) ~[starrocks-fe.jar:?]
    at com.starrocks.planner.IcebergMetadataScanNode.setupScanRangeLocations(IcebergMetadataScanNode.java:100) ~[starrocks-fe.jar:?]
    at com.starrocks.sql.plan.PlanFragmentBuilder$PhysicalPlanTranslator.visitPhysicalIcebergMetadataScan(PlanFragmentBuilder.java:1680) ~[starrocks-fe.jar:?]
    at com.starrocks.sql.plan.PlanFragmentBuilder$PhysicalPlanTranslator.visitPhysicalIcebergMetadataScan(PlanFragmentBuilder.java:461) ~[starrocks-fe.jar:?]
    at com.starrocks.sql.optimizer.operator.physical.PhysicalIcebergMetadataScanOperator.accept(PhysicalIcebergMetadataScanOperator.java:36) ~[starrocks-fe.jar:?]
    at com.starrocks.sql.plan.PlanFragmentBuilder$PhysicalPlanTranslator.visit(PlanFragmentBuilder.java:568) ~[starrocks-fe.jar:?]
    at com.starrocks.sql.plan.PlanFragmentBuilder$PhysicalPlanTranslator.translate(PlanFragmentBuilder.java:474) ~[starrocks-fe.jar:?]
    at com.starrocks.sql.plan.PlanFragmentBuilder.createPhysicalPlan(PlanFragmentBuilder.java:271) ~[starrocks-fe.jar:?]
    at com.starrocks.sql.StatementPlanner.createQueryPlan(StatementPlanner.java:338) ~[starrocks-fe.jar:?]
    at com.starrocks.sql.StatementPlanner.plan(StatementPlanner.java:150) ~[starrocks-fe.jar:?]
    at com.starrocks.connector.metadata.MetadataExecutor.asyncExecuteSQL(MetadataExecutor.java:40) ~[starrocks-fe.jar:?]
    at com.starrocks.connector.metadata.MetadataCollectJob.asyncCollectMetadata(MetadataCollectJob.java:68) ~[starrocks-fe.jar:?]
    at java.lang.Thread.run(Thread.java:840) ~[?:?]
Caused by: java.lang.NoSuchMethodException: Cannot find constructor for interface org.apache.iceberg.io.LocationProvider
    Missing com.company.CustomLocationProvider [java.lang.ClassNotFoundException: com.company.CustomLocationProvider]
    at org.apache.iceberg.common.DynConstructors.buildCheckedException(DynConstructors.java:252) ~[iceberg-common-1.10.0.jar:?]
    at org.apache.iceberg.common.DynConstructors$Builder.buildChecked(DynConstructors.java:210) ~[iceberg-common-1.10.0.jar:?]
    at org.apache.iceberg.LocationProviders.locationsFor(LocationProviders.java:50) ~[iceberg-core-1.10.0.jar:?]
    ... 18 more
    Suppressed: java.lang.ClassNotFoundException: com.company.CustomLocationProvider
        at jdk.internal.loader.BuiltinClassLoader.loadClass(BuiltinClassLoader.java:641) ~[?:?]
        at jdk.internal.loader.ClassLoaders$AppClassLoader.loadClass(ClassLoaders.java:188) ~[?:?]
        at java.lang.ClassLoader.loadClass(ClassLoader.java:525) ~[?:?]
        at java.lang.Class.forName0(Native Method) ~[?:?]
        at java.lang.Class.forName(Class.java:469) ~[?:?]
        at org.apache.iceberg.common.DynConstructors$Builder.classForName(DynConstructors.java:223) ~[iceberg-common-1.10.0.jar:?]
        at org.apache.iceberg.common.DynConstructors$Builder.impl(DynConstructors.java:145) ~[iceberg-common-1.10.0.jar:?]
        at org.apache.iceberg.LocationProviders.locationsFor(LocationProviders.java:49) ~[iceberg-core-1.10.0.jar:?]
        at org.apache.iceberg.BaseMetastoreTableOperations.locationProvider(BaseMetastoreTableOperations.java:234) ~[iceberg-core-1.10.0.jar:?]
        at org.apache.iceberg.BaseTable.locationProvider(BaseTable.java:255) ~[iceberg-core-1.10.0.jar:?]
        at com.starrocks.connector.share.iceberg.SerializableTable.<init>(SerializableTable.java:101) ~[starrocks-hadoop-ext.jar:?]
        at com.starrocks.connector.iceberg.IcebergMetadata.getSerializedMetaSpec(IcebergMetadata.java:634) ~[starrocks-fe.jar:?]
        at com.starrocks.connector.CatalogConnectorMetadata.getSerializedMetaSpec(CatalogConnectorMetadata.java:193) ~[starrocks-fe.jar:?]
        at com.starrocks.server.MetadataMgr.getSerializedMetaSpec(MetadataMgr.java:847) ~[starrocks-fe.jar:?]
        at com.starrocks.planner.IcebergMetadataScanNode.setupScanRangeLocations(IcebergMetadataScanNode.java:100) ~[starrocks-fe.jar:?]
        at com.starrocks.sql.plan.PlanFragmentBuilder$PhysicalPlanTranslator.visitPhysicalIcebergMetadataScan(PlanFragmentBuilder.java:1680) ~[starrocks-fe.jar:?]
        at com.starrocks.sql.plan.PlanFragmentBuilder$PhysicalPlanTranslator.visitPhysicalIcebergMetadataScan(PlanFragmentBuilder.java:461) ~[starrocks-fe.jar:?]
        at com.starrocks.sql.optimizer.operator.physical.PhysicalIcebergMetadataScanOperator.accept(PhysicalIcebergMetadataScanOperator.java:36) ~[starrocks-fe.jar:?]
        at com.starrocks.sql.plan.PlanFragmentBuilder$PhysicalPlanTranslator.visit(PlanFragmentBuilder.java:568) ~[starrocks-fe.jar:?]
        at com.starrocks.sql.plan.PlanFragmentBuilder$PhysicalPlanTranslator.translate(PlanFragmentBuilder.java:474) ~[starrocks-fe.jar:?]
        at com.starrocks.sql.plan.PlanFragmentBuilder.createPhysicalPlan(PlanFragmentBuilder.java:271) ~[starrocks-fe.jar:?]
        at com.starrocks.sql.StatementPlanner.createQueryPlan(StatementPlanner.java:338) ~[starrocks-fe.jar:?]
        at com.starrocks.sql.StatementPlanner.plan(StatementPlanner.java:150) ~[starrocks-fe.jar:?]
        at com.starrocks.connector.metadata.MetadataExecutor.asyncExecuteSQL(MetadataExecutor.java:40) ~[starrocks-fe.jar:?]
        at com.starrocks.connector.metadata.MetadataCollectJob.asyncCollectMetadata(MetadataCollectJob.java:68) ~[starrocks-fe.jar:?]
        at java.lang.Thread.run(Thread.java:840) ~[?:?]
```


### StarRocks version (Required)
 - You can get the StarRocks version by executing SQL `select current_version()`

4.0.10-ee

--- a/be/src/exprs/variant_functions.cpp
+++ b/be/src/exprs/variant_functions.cpp
@@ -78,8 +78,6 @@ Status VariantFunctions::variant_segments_prepare(FunctionContext* context, Func
     auto variant_path_status = VariantPathParser::parse(path_string);
     RETURN_IF(!variant_path_status.ok(), variant_path_status.status());
     context->set_function_state(scope, new VariantPath(std::move(variant_path_status.value())));
-    VLOG(10) << "Preloaded variant path: " << path_string;
-
     return Status::OK();
 }
 

--- a/java-extensions/hadoop-ext/src/main/java/com/starrocks/connector/share/iceberg/SerializableTable.java
+++ b/java-extensions/hadoop-ext/src/main/java/com/starrocks/connector/share/iceberg/SerializableTable.java
@@ -65,9 +65,6 @@ import java.util.Map;
 import java.util.UUID;
 
 public class SerializableTable implements Table, Serializable, HasTableOperations {
-
-    private static final long serialVersionUID = 1L;
-
     private final String name;
     private final String location;
     private final UUID uuid;

--- a/java-extensions/hadoop-ext/src/main/java/com/starrocks/connector/share/iceberg/SerializableTable.java
+++ b/java-extensions/hadoop-ext/src/main/java/com/starrocks/connector/share/iceberg/SerializableTable.java
@@ -24,6 +24,7 @@ import org.apache.iceberg.HasTableOperations;
 import org.apache.iceberg.HistoryEntry;
 import org.apache.iceberg.IncrementalAppendScan;
 import org.apache.iceberg.IncrementalChangelogScan;
+import org.apache.iceberg.LocationProviders;
 import org.apache.iceberg.ManageSnapshots;
 import org.apache.iceberg.OverwriteFiles;
 import org.apache.iceberg.PartitionSpec;
@@ -76,8 +77,9 @@ public class SerializableTable implements Table, Serializable, HasTableOperation
     private final String sortOrderAsJson;
     private final FileIO io;
     private final EncryptionManager encryption;
-    private final LocationProvider locationProvider;
     private final Map<String, SnapshotRef> refs;
+
+    private transient volatile LocationProvider lazyLocationProvider;
 
     private transient volatile Table lazyTable = null;
     private transient volatile Schema lazySchema = null;
@@ -98,7 +100,6 @@ public class SerializableTable implements Table, Serializable, HasTableOperation
         this.sortOrderAsJson = SortOrderParser.toJson(table.sortOrder());
         this.io = fileIO(fileIO);
         this.encryption = table.encryption();
-        this.locationProvider = table.locationProvider();
         this.refs = SerializableMap.copyOf(table.refs());
     }
 
@@ -129,7 +130,7 @@ public class SerializableTable implements Table, Serializable, HasTableOperation
                     }
 
                     TableOperations ops =
-                            new StaticTableOperations(metadataFileLocation, io, locationProvider);
+                            new StaticTableOperations(metadataFileLocation, io);
                     this.lazyTable = newTable(ops, name);
                 }
             }
@@ -239,7 +240,14 @@ public class SerializableTable implements Table, Serializable, HasTableOperation
 
     @Override
     public LocationProvider locationProvider() {
-        return locationProvider;
+        if (lazyLocationProvider == null) {
+            synchronized (this) {
+                if (lazyLocationProvider == null) {
+                    this.lazyLocationProvider = LocationProviders.locationsFor(location, properties);
+                }
+            }
+        }
+        return lazyLocationProvider;
     }
 
     @Override

--- a/java-extensions/hadoop-ext/src/main/java/com/starrocks/connector/share/iceberg/SerializableTable.java
+++ b/java-extensions/hadoop-ext/src/main/java/com/starrocks/connector/share/iceberg/SerializableTable.java
@@ -66,6 +66,8 @@ import java.util.UUID;
 
 public class SerializableTable implements Table, Serializable, HasTableOperations {
 
+    private static final long serialVersionUID = 1L;
+
     private final String name;
     private final String location;
     private final UUID uuid;

--- a/java-extensions/hadoop-ext/src/test/java/com/starrocks/connector/share/iceberg/SerializableTableTest.java
+++ b/java-extensions/hadoop-ext/src/test/java/com/starrocks/connector/share/iceberg/SerializableTableTest.java
@@ -1,0 +1,92 @@
+// Copyright 2021-present StarRocks, Inc. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.starrocks.connector.share.iceberg;
+
+import org.apache.iceberg.PartitionSpec;
+import org.apache.iceberg.Schema;
+import org.apache.iceberg.Table;
+import org.apache.iceberg.catalog.Namespace;
+import org.apache.iceberg.catalog.TableIdentifier;
+import org.apache.iceberg.inmemory.InMemoryCatalog;
+import org.apache.iceberg.inmemory.InMemoryFileIO;
+import org.apache.iceberg.types.Types;
+import org.junit.jupiter.api.Test;
+
+import java.util.HashMap;
+import java.util.Map;
+
+import static org.junit.jupiter.api.Assertions.assertDoesNotThrow;
+import static org.junit.jupiter.api.Assertions.assertNotNull;
+import static org.junit.jupiter.api.Assertions.assertThrows;
+
+public class SerializableTableTest {
+
+    private static final Schema SCHEMA = new Schema(
+            Types.NestedField.required(1, "id", Types.LongType.get()),
+            Types.NestedField.optional(2, "data", Types.StringType.get())
+    );
+
+    @Test
+    public void testConstructionSucceedsWithUnresolvableLocationProvider() {
+        // Regression test for https://github.com/StarRocks/starrocks/issues/73471
+        // SerializableTable should not eagerly load the LocationProvider at construction time.
+        // Custom LocationProvider JARs are write-path only and not available in FE classpath.
+        Map<String, String> props = new HashMap<>();
+        props.put("write.location-provider.impl", "com.nonexistent.CustomLocationProvider");
+
+        Table table = createTable(props);
+        InMemoryFileIO fileIO = new InMemoryFileIO();
+
+        // Should not throw ClassNotFoundException even though the provider class doesn't exist
+        SerializableTable serializable = assertDoesNotThrow(
+                () -> new SerializableTable(table, fileIO),
+                "SerializableTable construction must not trigger LocationProvider loading");
+
+        assertNotNull(serializable.schema());
+        assertNotNull(serializable.spec());
+        assertNotNull(serializable.refs());
+    }
+
+    @Test
+    public void testLocationProviderThrowsOnAccess() {
+        // When explicitly accessed, locationProvider() should throw because the class doesn't exist.
+        Map<String, String> props = new HashMap<>();
+        props.put("write.location-provider.impl", "com.nonexistent.CustomLocationProvider");
+
+        Table table = createTable(props);
+        SerializableTable serializable = new SerializableTable(table, new InMemoryFileIO());
+
+        assertThrows(IllegalArgumentException.class, serializable::locationProvider,
+                "locationProvider() should throw when the impl class is not in classpath");
+    }
+
+    @Test
+    public void testConstructionSucceedsWithNoCustomLocationProvider() {
+        Table table = createTable(new HashMap<>());
+        SerializableTable serializable = assertDoesNotThrow(
+                () -> new SerializableTable(table, new InMemoryFileIO()));
+
+        assertNotNull(serializable.schema());
+        assertNotNull(serializable.locationProvider());
+    }
+
+    private Table createTable(Map<String, String> properties) {
+        InMemoryCatalog catalog = new InMemoryCatalog();
+        catalog.initialize("test", new HashMap<>());
+        catalog.createNamespace(Namespace.of("db"));
+        TableIdentifier id = TableIdentifier.of("db", "tbl");
+        return catalog.createTable(id, SCHEMA, PartitionSpec.unpartitioned(), properties);
+    }
+}

